FirebaseInstallations/Source/Library/IIDMigration/FIRInstallationsIIDStore.m

```diff
@@ -0,0 +1,230 @@
+
+- (FBLPromise<NSString *> *)existingIID {
+  return [FBLPromise onQueue:dispatch_get_global_queue(QOS_CLASS_UTILITY, 0)
+                          do:^id _Nullable {
+                            if (![self hasPlistIIDFlag]) {
+                              return nil;
+                            }
+
+                            NSData *IIDPublicKeyData = [self IIDPublicKeyData];
+                            return [self IIDWithPublicKeyData:IIDPublicKeyData];
+                          }]
+      .validate(^BOOL(NSString *_Nullable IID) {
+        return IID.length > 0;
+      });
+}
+
+- (FBLPromise<NSNull *> *)deleteExistingIID {
+  return [FBLPromise onQueue:dispatch_get_global_queue(QOS_CLASS_UTILITY, 0)
+                          do:^id _Nullable {
+                            NSError *error;
+                            if (![self deleteIIDFlagFromPlist:&error]) {
+                              return error;
+                            }
+
+                            if (![self deleteIID:&error]) {
+                              return error;
+                            }
+
+                            return [NSNull null];
+                          }];
+}
+
+#pragma mark - IID decoding
+
+- (NSString *)IIDWithPublicKeyData:(NSData *)publicKeyData {
+  NSData *publicKeySHA1 = [self sha1WithData:publicKeyData];
+
+  const uint8_t *bytes = publicKeySHA1.bytes;
+  NSMutableData *identityData = [NSMutableData dataWithData:publicKeySHA1];
+
+  uint8_t b0 = bytes[0];
+  // Take the first byte and make the initial four 7 by initially making the initial 4 bits 0
+  // and then adding 0x70 to it.
+  b0 = 0x70 + (0xF & b0);
+  // failsafe should give you back b0 itself
+  b0 = (b0 & 0xFF);
+  [identityData replaceBytesInRange:NSMakeRange(0, 1) withBytes:&b0];
+  NSData *data = [identityData subdataWithRange:NSMakeRange(0, 8 * sizeof(Byte))];
+  return [self base64URLEncodedStringWithData:data];
+}
+
+- (NSData *)sha1WithData:(NSData *)data {
+  unsigned int outputLength = CC_SHA1_DIGEST_LENGTH;
+  unsigned char output[outputLength];
+  unsigned int length = (unsigned int)[data length];
+
+  CC_SHA1(data.bytes, length, output);
+  return [NSData dataWithBytes:output length:outputLength];
+}
+
+- (NSString *)base64URLEncodedStringWithData:(NSData *)data {
+  NSString *string = [data base64EncodedStringWithOptions:0];
+  string = [string stringByReplacingOccurrencesOfString:@"/" withString:@"_"];
+  string = [string stringByReplacingOccurrencesOfString:@"+" withString:@"-"];
+  string = [string stringByReplacingOccurrencesOfString:@"=" withString:@""];
+  return string;
+}
+
+#pragma mark - Keychain
+
+- (NSData *)IIDPublicKeyData {
+  NSString *tag = [self keychainKeyTagWithPrefix:kFIRInstallationsIIDKeyPairPublicTagPrefix];
+  NSDictionary *query = [self keyPairQueryWithTag:tag returnData:YES];
+
+  CFTypeRef keyRef = NULL;
+  OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)query, (CFTypeRef *)&keyRef);
```

Comment:
is this where the macOS start asking passcode?

Reply:
This one as well. I have this issue in `FIRSecureStorageTests` as well. That's why testing for macOS is disabled on Travis so far.